Mathlib/Topology/Spectral/Hom.lean

```diff
@@ -0,0 +1,227 @@
+  hf.isCompact_preimage_of_isOpen h₁ h₀
+#align is_compact.preimage_of_is_open IsCompact.preimage_of_isOpen
+
+theorem IsSpectralMap.continuous {f : α → β} (hf : IsSpectralMap f) : Continuous f :=
+  hf.toContinuous
+#align is_spectral_map.continuous IsSpectralMap.continuous
+
+theorem isSpectralMap_id : IsSpectralMap (@id α) :=
+  ⟨continuous_id, fun _s _ => id⟩
+#align is_spectral_map_id isSpectralMap_id
+
+theorem IsSpectralMap.comp {f : β → γ} {g : α → β} (hf : IsSpectralMap f) (hg : IsSpectralMap g) :
+    IsSpectralMap (f ∘ g) :=
+  ⟨hf.continuous.comp hg.continuous, fun _s hs₀ hs₁ =>
+    ((hs₁.preimage_of_isOpen hf hs₀).preimage_of_isOpen hg) (hs₀.preimage hf.continuous)⟩
+#align is_spectral_map.comp IsSpectralMap.comp
+
+end Unbundled
+
+/-- The type of spectral maps from `α` to `β`. -/
+structure SpectralMap (α β : Type _) [TopologicalSpace α] [TopologicalSpace β] where
+  /-- function between topological spaces-/
+  toFun : α → β
+  /-- proof that `toFun` is a spectral map-/
+  spectral' : IsSpectralMap toFun
+#align spectral_map SpectralMap
+
+section
+
+/-- `SpectralMapClass F α β` states that `F` is a type of spectral maps.
+
+You should extend this class when you extend `SpectralMap`. -/
+class SpectralMapClass (F : Type _) (α β : outParam <| Type _) [TopologicalSpace α]
+  [TopologicalSpace β] extends FunLike F α fun _ => β where
+  /-- statement that `F` is a type of spectal maps-/
+  map_spectral (f : F) : IsSpectralMap f
+#align spectral_map_class SpectralMapClass
+
+end
+
+export SpectralMapClass (map_spectral)
+
+attribute [simp] map_spectral
+
+-- See note [lower instance priority]
+-- porting note: `TopologicalSpace` params marked as implicit to address dangerous instances lint
+instance (priority := 100) SpectralMapClass.toContinuousMapClass {_ : TopologicalSpace α}
+    {_ : TopologicalSpace β} [SpectralMapClass F α β] : ContinuousMapClass F α β :=
+  { ‹SpectralMapClass F α β› with map_continuous := fun f => (map_spectral f).continuous }
+#align spectral_map_class.to_continuous_map_class SpectralMapClass.toContinuousMapClass
+
+instance [TopologicalSpace α] [TopologicalSpace β] [SpectralMapClass F α β] :
+    CoeTC F (SpectralMap α β) :=
+  ⟨fun f => ⟨_, map_spectral f⟩⟩
+
+/-! ### Spectral maps -/
+
+
+namespace SpectralMap
+
+variable [TopologicalSpace α] [TopologicalSpace β] [TopologicalSpace γ] [TopologicalSpace δ]
+
+/-- Reinterpret a `SpectralMap` as a `ContinuousMap`. -/
+def toContinuousMap (f : SpectralMap α β) : ContinuousMap α β :=
+  ⟨_, f.spectral'.continuous⟩
+#align spectral_map.to_continuous_map SpectralMap.toContinuousMap
+
+instance : SpectralMapClass (SpectralMap α β) α β
+    where
+  coe := SpectralMap.toFun
+  coe_injective' f g h := by cases f; cases g; congr
+  map_spectral f := f.spectral'
+
+/-- Helper instance for when there's too many metavariables to apply `fun_like.has_coe_to_fun`
+directly. -/
+instance : CoeFun (SpectralMap α β) fun _ => α → β :=
+  FunLike.hasCoeToFun
```

Comment:
```suggestion
-- Porting note: These CoeFun instances are not desirable in Lean 4.
--/-- Helper instance for when there's too many metavariables to apply `fun_like.has_coe_to_fun`
--directly. -/
--instance : CoeFun (SpectralMap α β) fun _ => α → β :=
--  FunLike.hasCoeToFun
```
We've been removing these direct coercion instances to ensure that the function coercion goes through `FunLike`.